ql/src/java/org/apache/hadoop/hive/ql/exec/HiveStructComparator.java

```diff
@@ -45,16 +48,14 @@ public int compare(Object key1, Object key2) {
         if (a1.size() == 0) {
             return 0;
         }
-        if (comparator == null) {
-            comparator = new WritableComparator[a1.size()];
-            // For struct all elements may not be of same type, so create comparator for each entry.
-            for (int i = 0; i < a1.size(); i++) {
-                comparator[i] = WritableComparatorFactory.get(a1.get(i), nullSafe, nullOrdering);
-            }
+        // For array, the length may not be fixed, so extend comparators on demand
+        for (int i = comparators.size(); i < a1.size(); i++) {
+            // For struct, all elements may not be of same type, so create comparator for each entry.
+            comparators.add(i, WritableComparatorFactory.get(a1.get(i), nullSafe, nullOrdering));
         }
         result = 0;
         for (int i = 0; i < a1.size(); i++) {
-            result = comparator[i].compare(a1.get(i), a2.get(i));
+            result = comparators.get(i).compare(a1.get(i), a2.get(i));
```

Comment:
For `ARRAY` comparisons we could even initialize and use only one comparator instead of creating the same one multiple times. If the size of the ARRAYs is small as in the example it will not make any real difference but for bigger ARRAYs we may feel both in terms of memory and CPU. Since this might need a bit of refactoring we can tackle it in a follow up JIRA, what do you think @okumin ?

Reply:
@zabetak Maybe, I have the same feeling. Basically, STRUCT and ARRAY are different data structures and we can have different approach. That would also make implementation straightforward.
My idea is let WritableComparatorFactory identify more precise types so that it can distinct STRUCT, ARRAY, and so on. It will require some effort since WritableComparatorFactory has to accept ObjectInspector or type information. But it sounds more robust than inferring data types from `Object`.
Anyway, I agree to create a follow-up ticket and I will do that if you have nothing more to discuss here.